### PR TITLE
HW 2 Submission S-A-Ismail

### PR DESCRIPTION
**Pull request by S-A-Ismail**

### CODE REVIEW

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_4.sql`: 
Hi,

Great job on your homework submission! You've done a good job of following the prompt and your code is well-structured and easy to read. Here are a few points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers most aspects of the homework prompt. However, the prompt asks for a group by on `user_id` and `browser_type`, but your code only groups by `user_id`. This means that your code does not account for different browser types for the same user. You can fix this by including `browser_type` in your `GROUP BY` clause, like so:

```sql
GROUP BY user_id, browser_type
```

2) **Logical Flow:** Your solution logically follows the given homework prompt. You've correctly implemented the three main transformations: unnesting the dates and converting them into powers of 2, summing those powers of 2 in a group by on `user_id`, and converting the sum to base 2.

3) **Readability:** Your code is quite readable. You've used meaningful variable names and your code is well-formatted. However, you could improve the readability of your code by adding comments explaining your thought process and what each part of the code does.

4) **Robustness and Error Handling:** Your code does not include any explicit error handling or checks for invalid inputs. While this may not be necessary for this specific assignment, it's a good practice to include error handling in your code to make it more robust.

5) **Modularity and Reusability:** Your code is modular and each part of the code is responsible for a specific task. This makes it easy to understand and reuse specific components in other contexts.

Overall, you've done a great job on this assignment! With a few minor adjustments, your code will be even better. Keep up the good work!

File: README.md
Comment (HjersmanJ) [inline README.md:1]: Try to have a README that summarizes the files included in this folder. It's part of the grading rubric and is recommended. Also include a descriptive PR comment that details the approach taken.

File: submission/query_1.sql [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        *,
+        ROW_NUMBER() OVER (
+            PARTITION BY game_id, team_id, player_id
+            ORDER BY game_id, team_id, player_id
+        ) AS rn
+    FROM bootcamp.nba_game_details
+)
+
+-- Select de-duplicated rows where rn = 1
+SELECT
```
Comment (HjersmanJ) [inline submission/query_1.sql:16]: SELECT * would suffice here.

File: submission/query_1.sql [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    SELECT
+        *,
+        ROW_NUMBER() OVER (
+            PARTITION BY game_id, team_id, player_id
+            ORDER BY game_id, team_id, player_id
+        ) AS rn
+    FROM bootcamp.nba_game_details
+)
+
+-- Select de-duplicated rows where rn = 1
```
Comment (HjersmanJ) [inline submission/query_1.sql:15]: This comment explains what the code does at too surface of a level. I'd rephrase this to say:

> We de-dup by selecting all but the first instance of a duplicated row (rn = 1).

**Comment by julie-scherer** [comment]
This is a LLM-generated comment for `query_4.sql`: 
Hi there,

Great job on your homework submission! You've done a good job of following the prompt and your code is well-structured and easy to read. Here are a few points of feedback:

1) **Coverage of the Homework Prompt:** Your code covers most aspects of the homework prompt. However, the prompt asks for a group by on `user_id` and `browser_type`, but your code only groups by `user_id`. This means that your code does not account for different browser types for the same user. You can fix this by including `browser_type` in your `GROUP BY` clause, like so:

```sql
GROUP BY user_id, browser_type
```

2) **Logical Flow:** Your solution logically follows the given homework prompt. You've correctly implemented the three main transformations: unnesting the dates and converting them into powers of 2, summing those powers of 2 in a group by on `user_id`, and converting the sum to base 2.

3) **Readability:** Your code is quite readable. You've used meaningful variable names and your code is well-formatted. However, you could improve the readability of your code by adding comments explaining your thought process and what each part of the code does.

4) **Robustness and Error Handling:** Your code does not include any explicit error handling or checks for invalid inputs. While this may not be necessary for this specific assignment, it's a good practice to include error handling in your code to make it more robust.

5) **Modularity and Reusability:** Your code is modular and each part of the code is responsible for a specific task. This makes it easy to understand and reuse specific components in other contexts.

Overall, you've done a great job on this assignment! With a few minor adjustments, your code will be even better. Keep up the good work!

File: submission/query_3.sql [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      browser_type,
+      "date",
+      COUNT(1)
+    FROM
+      reduced
+    WHERE
+      "date" = DATE('2023-01-03')
+    GROUP BY
+      user_id, browser_type,
+      "date"
```
Comment (HjersmanJ) [inline submission/query_3.sql:33]: [You can use GROUP BY aliasing as numbers.](https://trino.io/docs/current/sql/select.html#group-by-clause)

File: submission/query_4.sql [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+      saismail.user_devices_cumulated
+    WHERE
+      DATE = DATE('2023-01-07')
+  ),
+datelist_int AS (
+SELECT
+  user_id,
+  CAST(
+  SUM(
+  CASE
```
Comment (HjersmanJ) [inline submission/query_4.sql:15]: If you are performing a new line to allow a function (which is good), it's best to indent it for context readability.

**Review by HjersmanJ — APPROVED** [review]
You clearly understood the heart of the assignment. Just a few optimizations, formatting, and comment clarity feedback left. Good job!
